### PR TITLE
Another potential API design merging the existing API with some of Ralith's ideas

### PR DESCRIPTION
**Pull request by jimmycuadra**
Following on from #6. Like the title says, this is another attempt at a revised API. I'm still not happy with the API in #6 for two main reasons:

* It ties us to Hyper.
* Information about how to split request and response data into the various parts of the HTTP requests/responses lives in the implementation of the `Request` and `Response` trait, rather than the type system.

The API in this branch uses some of the ideas from #5 and #6 to augment the currently published API. The significant changes are:

* Associated types for all parts of the request (body, headers, path, and query) and all parts of the response (body, headers, status). Previously, headers were not accounted for, nor was the response status code.
* Moves many of the `Endpoint` methods into fields on the `Info` struct, like in #6.
* Provides an `Unused` struct that functions the same as the `Nothing` struct proposes in #5.
* The trait bounds on the associated types are very generic, such that they can be used to fill in the relevant part of the request/response when using Hyper, but does not tie the API to Hyper. Importantly, it also means that the implementing crate does not need to select an HTTP library, but crates consuming _that_ crate do. In other words, ruma-client-api would still be agnostic to Hyper, and all the Hyper-specific bindings would be in ruma-client.

Things I'm still not happy with:

* `Unused` sucks really badly. We really want specialization with intersecting implementations to represent the idea of optional associated types. See @withoutboats's [helpful replies to my question on /r/rust](https://www.reddit.com/r/rust/comments/5o89f6/optional_associated_types/dchmew3/).
* More associated types with more trait bounds means more code to implement for each endpoint. Not sure if this is avoidable.
* The need for separate `router_path` and `request_path` methods still feels clunky. It'd also be nice to find a way to replace `request_path` with a trait bound on `RequestPathParams` for symmetry with the other associated types.
* I'd really like the response to be represented with something more like this:
  ``` rust
  pub enum Response {
    AlreadyExists(AlreadyExistsHeaders, AlreadyExistsBody),
    Created(CreatedHeaders, CreatedBody),
  }
  ```
  That better represents the fact that the body will have a certain structure that changes based on what the response status code is. Right now the body of error responses (and other response bodies that vary based on status code) are not accounted for by the API. The problem is I'm not sure how to define an API like this with trait bounds. That is, we want this response type to be an enum where each variant has an associated status code, possibly headers, and de/serializable body struct. But every such enum might have a different number of variants, each with its own headers and body types, and AFAIK there isn't a way to use a variable number of type parameters.

Thoughts?

/cc @Ralith @vberger @eternaleye @jplatte @exul @neosam

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -97,38 +103,105 @@ pub enum Method {
...
 
-    /// Returns the HTTP method used by this endpoint.
-    fn method() -> Method;
+    /// Possible HTTP response headers.
+    type ResponseHeaders: TryFrom<&'a [u8]>;
 
-    /// Generates the path component of the URL for this endpoint using the supplied parameters.
-    fn request_path(params: Self::PathParams) -> String;
+    /// Possible HTTP response codes.
+    type ResponseStatus: Into<u16>;
```
Comment (farodin91) [inline src/lib.rs]: Why you use an Into instead of defining status? 
Status are limited in matrix.

Comment (jimmycuadra) [inline src/lib.rs]: Do you mean using an enum with only the status codes that are actually used by current API endpoints? If so, it'd be to be future-compatible with new endpoints that might use something else.

Comment (farodin91) [inline src/lib.rs]: Sorry for my slow reply. You implemented Into<u16> for an enum in the doc. If we implement one time in the library, we could use it everywhere.

Comment (jimmycuadra) [inline src/lib.rs]: Ah, I see what you mean. The idea was that each endpoint may have its own semantics for what its various response status codes mean. They all map to `u16`s, but don't all mean the same thing. For example, a 200 OK might mean "resource created" for one endpoint and "the user has been authenticated" in another.

**Comment by jimmycuadra** [comment]
Closing in favor of the API I just landed on master.

**Pull request closed without merging**